src/devtools/store.js

```diff
@@ -425,22 +448,67 @@ export default class Store extends EventEmitter {
     // We do this to avoid mismatches on e.g. CommitTreeBuilder that would cause errors.
   }
 
+  // TODO Maybe split this into two methods: expand() and collapse()
   toggleIsCollapsed(id: number, isCollapsed: boolean): void {
     const element = this.getElementByID(id);
     if (element !== null) {
-      const oldWeight = element.isCollapsed ? 1 : element.weight;
-      element.isCollapsed = isCollapsed;
-      const newWeight = element.isCollapsed ? 1 : element.weight;
-      const weightDelta = newWeight - oldWeight;
+      if (isCollapsed) {
+        if (element.type === ElementTypeRoot) {
+          throw Error('Root nodes cannot be collapsed');
+        }
 
-      this._weightAcrossRoots += weightDelta;
+        if (element.isCollapsed) {
+          return;
+        }
 
-      let parentElement = this._idToElement.get(element.parentID);
-      while (parentElement != null) {
-        parentElement.weight += weightDelta;
-        parentElement = this._idToElement.get(parentElement.parentID);
+        element.isCollapsed = true;
+
+        const weightDelta = 1 - element.weight;
+
+        let parentElement = ((this._idToElement.get(
+          element.parentID
+        ): any): Element);
+        while (parentElement != null) {
+          parentElement.weight += weightDelta;
+          parentElement = this._idToElement.get(parentElement.parentID);
+        }
+      } else {
+        let currentElement = element;
+        while (currentElement != null) {
+          const oldWeight = currentElement.isCollapsed
+            ? 1
+            : currentElement.weight;
+          currentElement.isCollapsed = false;
+          const newWeight = currentElement.isCollapsed
+            ? 1
+            : currentElement.weight;
+          const weightDelta = newWeight - oldWeight;
+
+          let parentElement = ((this._idToElement.get(
+            currentElement.parentID
+          ): any): Element);
+          while (parentElement != null) {
+            parentElement.weight += weightDelta;
+            if (parentElement.isCollapsed) {
```

Comment:
It's important to bail out on collapsed parents here, because we are going to walk up the tree and expand them too (and we don't want to add their weight multiple times).